### ISSUE TITLE
Add support for dynamically repairing Workspace One

### ISSUE THREAD

**Issue by w0de** [uber/cpe-chef-cookbooks#26]
Recently, I challenge a support engineer to break our management tools on his Mac.

He succeeded - and the only tool that could not be easily repaired from a basis of MDM enrollment was, surprisingly, `hubcli`. Workspace One continues to function without it.

Since this cookbook already supports installing the agent, and is smart about the existence of `hubcli`, I think it would be neat if it provided something like this:

```
node['cpe_workspaceone']['remediate_missing_hubcli'] = true
```

If true, cpe_workspaceone would reinstall the package. This also leaves the door open for sophisticated health checks.

**Comment by w0de** [comment]
Implementation could be as simple as force-forgetting the package receipt for the agent if hubcli is missing.

**Comment by erikng** [comment]
We should probably just do this automatically. If an admin wants to install it, chances are they want a desired state (for it to actually work). 

What detection logic do we need?

**Comment by erikng** [comment]
Just if hubcli is missing?

**Comment by w0de** [comment]
I think as simple as `ws1_forget_package! if ws1_hubcli_missing` works. 

But I'm not sure if I want this to happen all the time. I can be timid about installing packages in the background - what if there's a logic error and we loop forgetting/installing the package?

**Comment by w0de** [comment]
Also while we're here I think the `ws1_` should be refactored to be a class. I'll open a separate issue/PR if you're interested. Ie:

```ruby
class Chef
  class Node
    def ws1
      @ws1 ||= WorkspaceOneUtils.new(self)
    end

    class WorkspaceOneUtils
      def new(node_object)
        @node = node_object
      end

      def node
        @node
      end

      def device_attributes
        ...
      end
    end
  end
end

node.ws1.device_attributes
```

**Comment by natewalck** [comment]
I think if ws1 becomes a class, it should be *outside* of `node`.

**Comment by w0de** [comment]
Yes @natewalck I was just saying to myself that I'm making `Node` a god class.

Going down the route, one might end up accidentally managing machine state through the node object, instead of through chef resources.

**Comment by natewalck** [comment]
Another fun discussion is if we should *ever* be overloading `node` with custom classes or if we should break it all out into a `CPE` class. :D

**Comment by w0de** [comment]
I think hewing to an established chef patten would be ideal (I feel that `cpe_profiles` and `cpe_launchd` have flawed designs). 

Starting here: https://blog.chef.io/writing-libraries-in-chef-cookbooks/ -

```ruby
# Not a fan of "CPE" as a module name though
module CPE
  module WorkspaceOne

    include Chef::Mixin::ShellOut

    def hubcli
      cmd = shell_out!("#{hubcli_path}", {:returns => [0]})
    end
  end
end

# Not sure of resource name
Chef::Resource::cpe_workspaceone.send(:include, CPE::WorkspaceOne)
```

**Comment by natewalck** [comment]
I agree. One thing that we need to test out is can a class access node by default? I believe that answer is no...so if we want to use node data for anything, we'll have to figure out how to get it from the class. I'm sure there must be some way to do this via subclassing or some such.

**Comment by w0de** [comment]
This is something that I'm playing with for an internal recipe:

```ruby
include Chef::Mixin::ShellOut

module EngineeringHelpers
  def self.sdk_path
    shell_out!("/usr/bin/xcrun --show-sdk-path").stdout.strip
  end
end

class Chef
  class Resource
    def eng_helpers
      EngineeringHelpers
    end
  end
  class Recipe
    def eng_helpers
      EngineeringHelpers
    end
  end
end
```

```ruby
# recipes/default.rb
eng_helpers.sdk_path
```

**Comment by erikng** [comment]
@natewalck can you put the memoizing feature in this cookbook for our next sprint so we can get some traction on this diff?
